zebra-network/src/peer/server.rs

```diff
@@ -82,66 +91,60 @@ where
-                            //Some(Err(e)) => self.fail_with(e.into()),
-                            // XXX remove this when we parse all message types
-                            Some(Err(e)) => {
-                                error!(%e);
-                            }
-                            // We got a peer message and it was well-formed.
-                            Some(Ok(msg)) => self.handle_message_as_request(msg).await,
+                        // XXX switch back to hard failure when we parse all message types
+                        //Either::Left((Some(Err(e)), _)) => self.fail_with(e.into()),
+                        Either::Left((Some(Err(e)), _)) => error!(%e),
+                        Either::Left((Some(Ok(msg)), _)) => {
+                            self.handle_message_as_request(msg).await
                         }
+                        Either::Right((None, _)) => {
+                            info!("client stream closed, shutting down");
+                            return;
+                        }
+                        Either::Right((Some(req), _)) => self.handle_client_request(req).await,
                     }
-                },
+                }
                 // We're awaiting a response to a client request,
-                // so only listen to peer messages, not further requests.
+                // so wait on either a peer message, or on a request timeout.
                 ServerState::AwaitingResponse { .. } => {
-                    let msg = peer_rx_fut.await;
-                    peer_rx_fut = peer_rx.next().fuse();
-                    match msg {
-                        // The peer channel has closed -- no more messages.
-                        // However, we still need to flush pending client requests.
-                        None => self.fail_with(format_err!("peer closed connection").into()),
-                        // We got a peer message but it was malformed.
-                        //Some(Err(e)) => self.fail_with(e.into()),
-                        // XXX remove this when we parse all message types
-                        Some(Err(e)) => {
-                            error!(%e);
+                    trace!("awaiting response to client request");
+                    let timer_ref = self
+                        .request_timer
+                        .as_mut()
+                        .expect("timeout must be set while awaiting response");
+                    match future::select(peer_rx.next(), timer_ref).await {
+                        Either::Left((None, _)) => {
+                            self.fail_with(format_err!("peer closed connection").into())
+                        }
+                        // XXX switch back to hard failure when we parse all message types
+                        //Either::Left((Some(Err(e)), _)) => self.fail_with(e.into()),
+                        Either::Left((Some(Err(e)), _)) => error!(%e),
```

Comment:
```suggestion
                        Either::Left((Some(Err(peer_err)), _timer)) => error!(%e),
```
see discussion above